rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,334 @@
+
+#### Communication Architecture
+
+The helper and daemon will now exchange messages over a variety of data streams, allowing each process to prioritize data streams differently. Correctly optimizing this prioritization on the daemon side is important, since OCaml is single threaded by nature (for now). In particular, the daemon needs to be able to priotize processing and validating certain network messages in order to ensure that the node keeps in sync with the network and forwards relevant information for others to stay in sync. Specifically, the daemon needs to prioritize processing and validating new block gossips so that they can be forwarded to other nodes on the network in a timely manor.
+
+The transport layer we will use for these data streams will be Unix pipes. The parent daemon process can create all the required Unix pipes for the various data streams, and pass the correct file descriptors (either the write or read descriptors depending on the direction of the pipe) to the child helper process when it initializes. Pipes are considered preferable over shared memory for the data streams since they already provide synchronization primitives for reading/writing and are easier to implement correctly, though shared memory would be likely be slightly more optimized.
+
+Below is a proposed set of data streams we would setup for the helper IPC protocol. Keep in mind that some of these pipes cannot be setup without helper message type awareness, and thus will not be available separately until we perfrom a hard-fork update.
+
+- stdin (used only for initialization message, then closed)
+- stdout (used only for helper logging)
+- stderr (used only for helper logging)
+- stats\_in (publishes helper stats to daemon on an interval)
+- block\_gossip\_in (incoming block gossip messages)
+- mempool\_gossip\_in (other incoming gossip messages, related to mempool state)
+- response\_in (incoming RPC responses)
+- request\_in (incoming RPC requests)
+- validation\_out (all validations except request validations, which are bundled with responses)
+- data\_out (responses, broadcasts)
+
+The rough priorities for reading the incoming pipes from the daemon process would be:
+
+- block\_gossip\_in
+- response\_in
+- request\_in
+- mempool\_gossip\_in
```

Comment:
My thoughts on this are that this concern will be mostly addressed by bitswap (I believe) since the broadcast messages will be very small once we are in that world. Doing this pipe read prioritization correctly is rather complicated, so I would like to avoid splitting this up and adding additional complexity to that system if possible. Do you believe that the vastly reduced broadcast message sizes in bitswap + rate limiting in trust scoring address this concern, or do you think we need to still put more thought into splitting up these pipes?